### PR TITLE
General bug fixes to function calling

### PR DESCRIPTION
**Pull request by enigmarikki**
Bug fixes: 
1. Fixed the dependency bug
2. Fixed fetching data from subaccounts
New features:
1. Enabled leverage trading on perps
2. Improved and added docker setup

## Release Notes

- **New Features**
	- Introduced new methods for auction data retrieval: `fetch_auctions`, `fetch_latest_auction`, and `fetch_auction_bids`.
	- Enhanced order placement methods with a new `leverage` parameter for derivatives.
	- Added a utility function for base64 conversion.

- **Improvements**
	- Improved handling of deposit data with updated method for `get_subaccount_deposits`.
	- Increased API request timeout for better response handling.
	- Enhanced error handling in various methods for more robust responses.
	- Transitioned several methods to asynchronous execution for improved performance.

- **Documentation**
	- Updated README with installation options for local and Docker setups, along with usage instructions.

- **Dependencies**
	- Added `quart` and `pyyaml` to requirements.
	- Updated `OPENAI_API_KEY` placeholder in `compose.yaml` for improved security.

### CODE REVIEW

File: injective_functions/utils/initializers.py [hunk 1]
```diff
@@ -40,7 +40,7 @@ async def init_client(self):
     async def build_and_broadcast_tx(self, msg):
         """Common function to build and broadcast transactions"""
         try:
-            self.init_client()
+            await self.init_client()
```
Comment (defser) [inline injective_functions/utils/initializers.py:43]: i didn't want to commit this in my PR, glad you got it.

File: injective_functions/auction/__init__.py [hunk 2]
```diff
@@ -13,10 +16,35 @@ def __init__(self, chain_client) -> None:
...
         return await self.chain_client.build_and_broadcast_tx(msg)
+
+    async def fetch_auctions(self) -> Dict:
+        try:
+            auctions = await self.chain_client.client.fetch_auctions()
+            return {"success": True, "result": auctions["auctions"]}
+        except Exception as e:
+            return {"success": False, "error": detailed_exception_info(e)}
+
+    async def fetch_latest_auction(self) -> Dict:
```
Comment (djosse) [inline injective_functions/auction/__init__.py:34]: I'd add a little more safety to this.. something like.. (and ensuring that we return an error on the Exception case)
  ```
try:
        result = await self.fetch_auctions()
        
        # Ensure the result structure is as expected
        if result.get("success") and result.get("result"):
            latest_auction = result["result"][-1]  # Safely access the latest auction
            return {"success": True, "result": latest_auction}
        
except Exception as e:
        return {"success": False, "error": f"Unexpected error: {str(e)}"}
```

File: injective_functions/exchange/exchange.py [hunk 3]
```diff
@@ -19,29 +19,28 @@ def __init__(self, chain_client) -> None:
...
-        self, subaccount_idx: int, denoms: str = None
+        self, subaccount_idx: int, denoms: List[str] = None
     ) -> Dict:
         try:
 
-            subaccount_id = await self.chain_client.address.get_subaccount_id(
-                subaccount_idx
-            )
+            subaccount_id = self.chain_client.address.get_subaccount_id(subaccount_idx)
             deposits = await self.chain_client.client.fetch_subaccount_deposits(
```
Comment (djosse) [inline injective_functions/exchange/exchange.py:27]: Don't love the pattern of reusing variables.. would refactor: 
```
        deposits_response = await self.chain_client.client.fetch_subaccount_deposits(subaccount_id=subaccount_id)
        deposits = deposits_response.get("deposits", {})
```

File: injective_functions/exchange/exchange.py [hunk 4]
```diff
@@ -19,29 +19,28 @@ def __init__(self, chain_client) -> None:
...
                 subaccount_id=subaccount_id
-            )["deposits"]
+            )
+            deposits = deposits["deposits"]
             denom_decimals = await fetch_decimal_denoms(self.chain_client.network_type)
             human_readable_deposits = {}
-            if len(denoms) > 0:
+            if denoms:
```
Comment (djosse) [inline injective_functions/exchange/exchange.py:33]: This block looks somewhat difficult to parse, this logic is way more cleaner: 

```
     # Determine target denoms (all if not specified)
        target_denoms = denoms or deposits.keys()

        for denom in target_denoms:
            if denom in deposits and denom in denom_decimals:
                # Convert balances to human-readable format
                available_balance = int(deposits[denom]["availableBalance"]) / 10 ** denom_decimals[denom]
                total_balance = int(deposits[denom]["totalBalance"]) / 10 ** denom_decimals[denom]
                human_readable_deposits[denom] = {
                    "available_balance": f"{available_balance:.6f}",
                    "total_balance": f"{total_balance:.6f}",
                }
            else:
                # Handle any missing balances
                human_readable_deposits[denom] = {
                    "available_balance": "balance not found",
                    "total_balance": "balance not found",
                }
```

File: injective_functions/exchange/exchange.py [hunk 5]
```diff
@@ -121,7 +123,7 @@ async def get_mid_price_and_tob_derivatives_market(self, market_id: str) -> Dict
             )
             return {"success": True, "result": res}
         except Exception as e:
-            return {"success": False, "result": detailed_exception_info(e)}
+            return {"success": False, "error": detailed_exception_info(e)}
```
Comment (djosse) [inline injective_functions/exchange/exchange.py:126]: Good! exceptions should return error

File: injective_functions/exchange/exchange.py [hunk 6]
```diff
@@ -19,29 +19,28 @@ def __init__(self, chain_client) -> None:
         super().__init__(chain_client)
 
     async def get_subaccount_deposits(
-        self, subaccount_idx: int, denoms: str = None
+        self, subaccount_idx: int, denoms: List[str] = None
```
Comment (djosse) [inline injective_functions/exchange/exchange.py:22]: Denoms can be optional yes? Let's mark it as such

File: injective_functions/auction/__init__.py [hunk 7]
```diff
@@ -1,6 +1,9 @@
 from decimal import Decimal
 from injective_functions.base import InjectiveBase
 from typing import Dict
+from injective_functions.utils.helpers import (
```
Comment (djosse) [inline injective_functions/auction/__init__.py:4]: I like to order imports from more generic to more specific.. so 
```
from decimal import Decimal
from typing import Dict
from injective_functions.base import InjectiveBase
from injective_functions.utils.helpers import (
```

File: injective_functions/exchange/trader.py [hunk 8]
```diff
@@ -150,8 +156,7 @@ async def place_spot_market_order(
     async def cancel_spot_limit_order(
         self, market_id: str, subaccount_idx: int, order_hash: str
     ):
-        await self.chain_client.init_client()
-
+        order_hash = base64convert(order_hash)
```
Comment (djosse) [inline injective_functions/exchange/trader.py:159]: Do not rewrite variables. Suggest using new names for new variables, will make this easier to read: 
```
converted_order_hash = base64.. 
resolved_market_id = await.. 
```

File: readme.md [hunk 9]
```diff
@@ -90,4 +130,4 @@ Contributions are welcome! Please fork the repository and submit a pull request.
...
...e nor shall it be construed as offering legal, financial, investment, or business advice. No warranties or guarantees are made about the accuracy of any information or code contained herein. The legal and regulatory risks inherent to digital assets are not the subject of this Repository, and no decision to buy, sell, exchange, or otherwise utilize any digital asset is recommended based on the content of this Repository. USE AT YOUR OWN RISK. THE AUTHORS AND ALL AFFILIATES ASSUME NO RESPONSIBILITY FOR YOUR TRADING RESULTS. For guidance regarding the possibility of said risks, one should consult with his or her own appropriate legal and/or regulatory counsel. One should consult one's own advisors for any and all matters related to finance, business, legal, regulatory, and technical knowledge or expertise. Use, fork or publication of this Repository represents an acknowledgment that Anti Social Social Capital Ltd and Injective Labs Inc. are indemnified from any form of liability associated with actions taken by any other interested party. That explicit indemnification is acknowledged by the user or publisher to be legally binding and severable from all other portions of this document.
```
Comment (djosse) [inline readme.md:133]: What is this "Anti Social Social Capital"?

File: injective_functions/auction/__init__.py (hunk 7 above)
Comment (enigmarikki) [reply injective_functions/auction/__init__.py:4]: I used a linting tool to automatically manage all imports

File: readme.md (hunk 9 above)
Comment (enigmarikki) [reply readme.md:133]: Good catch need to remove it

File: injective_functions/exchange/exchange.py (hunk 6 above)
Comment (enigmarikki) [reply injective_functions/exchange/exchange.py:22]: weird, i did initally set that to none

File: injective_functions/auction/__init__.py (hunk 2 above)
Comment (enigmarikki) [reply injective_functions/auction/__init__.py:34]: uses standardized exceptions to give us detailed info on errors, so all the functions either return {"success": True, "result": Dict} or {"sucess": False, "result": Dict} this ensures the gpt function calling easily reads through errors

File: injective_functions/exchange/exchange.py (hunk 4 above)
Comment (enigmarikki) [reply injective_functions/exchange/exchange.py:33]: added comments and refactor

**Review by djosse — APPROVED** [review]

**Pull request merged**